central/apitoken/expiration/expiration_test.go

```diff
@@ -0,0 +1,134 @@
+	"github.com/stackrox/rox/pkg/uuid"
+	"github.com/stretchr/testify/suite"
+)
+
+func TestAPITokenExpirationNotifier(t *testing.T) {
+	suite.Run(t, new(apiTokenExpirationNotifierTestSuite))
+}
+
+type apiTokenExpirationNotifierTestSuite struct {
+	suite.Suite
+
+	testpostgres *pgtest.TestPostgres
+
+	datastore apiTokenDataStore.DataStore
+	notifier  *expirationNotifierImpl
+}
+
+func (s *apiTokenExpirationNotifierTestSuite) SetupSuite() {
+	if !env.PostgresDatastoreEnabled.BooleanSetting() {
+		s.T().Skip("Notification of expired API tokens is only supported in Postgres mode.")
+	}
+}
+
+func (s *apiTokenExpirationNotifierTestSuite) SetupTest() {
+	s.testpostgres = pgtest.ForT(s.T())
+	s.datastore = apiTokenDataStore.NewPostgres(s.testpostgres.DB)
+	s.notifier = newExpirationNotifier(s.datastore)
+}
+
+func (s *apiTokenExpirationNotifierTestSuite) TearDownTest() {
+	s.testpostgres.Teardown(s.T())
+}
+
+func truncateToMicroSeconds(timestamp *types.Timestamp) *types.Timestamp {
+	outputTs := timestamp.Clone()
+	outputTs.Nanos = 1000 * (outputTs.Nanos / 1000)
+	return outputTs
+}
+
+func generateToken(now *time.Time,
+	expiration *time.Time,
+	revoked bool) *storage.TokenMetadata {
+	var protoNow *types.Timestamp
+	var protoExpiration *types.Timestamp
+	if now != nil {
+		protoNow = protoconv.ConvertTimeToTimestamp(*now)
+		protoNow = truncateToMicroSeconds(protoNow)
+	}
+	if expiration != nil {
+		protoExpiration = protoconv.ConvertTimeToTimestamp(*expiration)
+		protoExpiration = truncateToMicroSeconds(protoExpiration)
+	}
+	return &storage.TokenMetadata{
+		Id:         uuid.NewV4().String(),
+		Name:       "Generated Test Token",
+		Roles:      []string{"Admin"},
+		IssuedAt:   protoNow,
+		Expiration: protoExpiration,
+		Revoked:    revoked,
+	}
+}
+
+func (s *apiTokenExpirationNotifierTestSuite) TestSelectTokenAboutToExpire() {
+	ctx := sac.WithAllAccess(context.Background())
+	now := time.Now()
+	expiration := now.Add(2 * time.Hour)
+	expiresUntil := now.Add(5 * time.Hour)
+	token := generateToken(&now, &expiration, false)
+	s.Require().NoError(s.datastore.AddToken(ctx, token))
+
+	fetchedTokens, err := s.notifier.listItemsToNotify(now, expiresUntil)
+	s.NoError(err)
+	expectedResult := token
+	expectedResults := []*storage.TokenMetadata{expectedResult}
+	s.ElementsMatch(expectedResults, fetchedTokens)
+}
+
+func (s *apiTokenExpirationNotifierTestSuite) TestDontSelectTokenNotAboutToExpire() {
+	ctx := sac.WithAllAccess(context.Background())
+	now := time.Now()
+	expiration := now.Add(7 * time.Hour)
+	expiresUntil := now.Add(5 * time.Hour)
+	token := generateToken(&now, &expiration, false)
+	s.Require().NoError(s.datastore.AddToken(ctx, token))
+
+	fetchedTokens, err := s.notifier.listItemsToNotify(now, expiresUntil)
+	s.NoError(err)
+	expectedResults := []*storage.TokenMetadata{}
+	s.ElementsMatch(expectedResults, fetchedTokens)
+}
+
+func (s *apiTokenExpirationNotifierTestSuite) TestDontSelectRevokedToken() {
```

Comment:
```
=== RUN   TestAPITokenExpirationNotifier/TestDontSelectRevokedToken
    expiration_test.go:119: 
        	Error Trace:	/home/janisz/go/src/github.com/stackrox/stackrox/central/apitoken/expiration/expiration_test.go:119
        	Error:      	elements differ
        	            	
        	            	extra elements in list B:
        	            	([]interface {}) (len=1) {
        	            	 (*storage.TokenMetadata)({
        	            	  Id: (string) (len=36) "08cb277d-33b4-470c-a1f4-9624fc149936",
        	            	  Name: (string) (len=20) "Generated Test Token",
        	            	  Roles: ([]string) (len=1) {
        	            	   (string) (len=5) "Admin"
        	            	  },
        	            	  IssuedAt: (*types.Timestamp)({
        	            	   Seconds: (int64) 1677679569,
        	            	   Nanos: (int32) 702390000,
        	            	   XXX_NoUnkeyedLiteral: (struct {}) {
        	            	   },
        	            	   XXX_unrecognized: ([]uint8) <nil>,
        	            	   XXX_sizecache: (int32) 0
        	            	  }),
        	            	  Expiration: (*types.Timestamp)({
        	            	   Seconds: (int64) 1677686769,
        	            	   Nanos: (int32) 702390000,
        	            	   XXX_NoUnkeyedLiteral: (struct {}) {
        	            	   },
        	            	   XXX_unrecognized: ([]uint8) <nil>,
        	            	   XXX_sizecache: (int32) 0
        	            	  }),
        	            	  Revoked: (bool) true,
        	            	  Role: (string) "",
        	            	  XXX_NoUnkeyedLiteral: (struct {}) {
        	            	  },
        	            	  XXX_unrecognized: ([]uint8) <nil>,
        	            	  XXX_sizecache: (int32) 0
        	            	 })
        	            	}
        	            	
        	            	
        	            	listA:
        	            	([]*storage.TokenMetadata) {
        	            	}
        	            	
        	            	
        	            	listB:
        	            	([]*storage.TokenMetadata) (len=1) {
        	            	 (*storage.TokenMetadata)({
        	            	  Id: (string) (len=36) "08cb277d-33b4-470c-a1f4-9624fc149936",
        	            	  Name: (string) (len=20) "Generated Test Token",
        	            	  Roles: ([]string) (len=1) {
        	            	   (string) (len=5) "Admin"
        	            	  },
        	            	  IssuedAt: (*types.Timestamp)({
        	            	   Seconds: (int64) 1677679569,
        	            	   Nanos: (int32) 702390000,
        	            	   XXX_NoUnkeyedLiteral: (struct {}) {
        	            	   },
        	            	   XXX_unrecognized: ([]uint8) <nil>,
        	            	   XXX_sizecache: (int32) 0
        	            	  }),
        	            	  Expiration: (*types.Timestamp)({
        	            	   Seconds: (int64) 1677686769,
        	            	   Nanos: (int32) 702390000,
        	            	   XXX_NoUnkeyedLiteral: (struct {}) {
        	            	   },
        	            	   XXX_unrecognized: ([]uint8) <nil>,
        	            	   XXX_sizecache: (int32) 0
        	            	  }),
        	            	  Revoked: (bool) true,
        	            	  Role: (string) "",
        	            	  XXX_NoUnkeyedLiteral: (struct {}) {
        	            	  },
        	            	  XXX_unrecognized: ([]uint8) <nil>,
        	            	  XXX_sizecache: (int32) 0
        	            	 })
        	            	}
        	Test:       	TestAPITokenExpirationNotifier/TestDontSelectRevokedToken
    --- FAIL: TestAPITokenExpirationNotifier/TestDontSelectRevokedToken (2.17s)
```

Reply:
@dhaus67 spotted the source of this issue in https://github.com/stackrox/stackrox/pull/4968 (bad search field in proto annotation).
This should be fixed now.